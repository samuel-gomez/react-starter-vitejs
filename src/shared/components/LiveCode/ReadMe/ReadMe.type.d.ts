import { CodeMarkdown } from './ReadMe';
import { TReturnUseGithubReadme } from './ReadMe.hook';

export type TReadMe = {
  markdownContent: TReturnUseGithubReadme['markdownContent'];
  isFetching: TReturnUseGithubReadme['isFetching'];
  error: TReturnUseGithubReadme['error'];
  refetch: TReturnUseGithubReadme['refetch'];
  CodeMarkdownFn?: typeof CodeMarkdown;
};

export type THighlightWrapper = {
  className: string;
  style: object;
  child: React.ReactNode;
  getLineProps: (input: LineInputProps) => LineOutputProps;
  getTokenProps: (input: TokenInputProps) => TokenOutputProps;
  tokens: Token[][];
};

export type LineInputProps = {
  key?: React.Key;
  style?: CSSProperties;
  className?: string;
  line: Token[];
};

export type TokenOutputProps = {
  key?: React.Key;
  style?: CSSProperties;
  className: string;
  children: string;
};

export type TokenInputProps = {
  key?: React.Key;
  style?: CSSProperties;
  className?: string;
  token: Token;
};

export type Token = {
  types: string[];
  content: string;
  empty?: boolean;
};

export type LineOutputProps = {
  key?: React.Key;
  style?: CSSProperties;
  className: string;
};
