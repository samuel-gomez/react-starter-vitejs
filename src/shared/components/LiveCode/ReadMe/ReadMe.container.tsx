/* eslint-disable react/no-unused-prop-types */
import { useGithubReadme } from './ReadMe.hook';
import ReadMe from './ReadMe';

type TReadMeContainer = {
  hideComponent?: boolean;
  title?: string;
  icon?: string;
  githubPackage?: string;
  useGithubReadmeFn?: typeof useGithubReadme;
};

const ReadMeContainer = ({ githubPackage, useGithubReadmeFn = useGithubReadme }: TReadMeContainer) => {
  const { markdownContent, isFetching, error, refetch } = useGithubReadmeFn({ githubPackage });

  return <ReadMe markdownContent={markdownContent} isFetching={isFetching} error={error} refetch={refetch} />;
};

export default ReadMeContainer;
