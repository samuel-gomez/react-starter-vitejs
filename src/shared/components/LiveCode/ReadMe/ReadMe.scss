.markdown-body {
  min-height: 300px;
  padding: 2rem;
}
.markdown-body .highlight pre,
.markdown-body pre {
  padding: 0;
}
