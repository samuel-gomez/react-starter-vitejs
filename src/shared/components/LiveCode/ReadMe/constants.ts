const styleHighlight = { padding: '1rem', fontSize: '0.8rem', position: 'relative' } as const;

export default styleHighlight;
