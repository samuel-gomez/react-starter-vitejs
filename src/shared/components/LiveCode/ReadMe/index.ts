export { default } from './ReadMe.container';
