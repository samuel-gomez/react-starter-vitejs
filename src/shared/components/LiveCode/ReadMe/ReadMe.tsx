import ReactMarkdown from 'react-markdown';
import remarkGfm from 'remark-gfm';
import rehypeSlug from 'rehype-slug';
import Highlight, { defaultProps } from 'prism-react-renderer';
import vsDark from 'prism-react-renderer/themes/vsDark';
import { createId } from '@axa-fr/react-toolkit-core/dist/esm/useId';
import 'github-markdown-css/github-markdown-light.css';
import Resilience from 'shared/components/Resilience';
import { ReactNode } from 'react';
import Loader, { setLoaderMode } from 'shared/components/Loader';
import { Tanomaly } from 'shared/types';
import ClipBoard from '../ClipBoard';
import type { THighlightWrapper, Token, TReadMe } from './ReadMe.type';
import styleHighlight from './constants';
import './ReadMe.scss';

export const regexLanguage = (className: string | undefined) => /language-(\w+)/.exec(className || '');

const HighlightWrapper = ({ className, style, child, getLineProps, getTokenProps, tokens }: THighlightWrapper) => (
  <pre className={className} style={{ ...style, ...styleHighlight }}>
    <ClipBoard content={String(child).replace(/\n$/, '')} />
    {tokens.map((line: Token[], i: number) => {
      const keyLine = createId();
      return (
        <div key={keyLine} {...getLineProps({ line, key: keyLine })}>
          <span className="token-line__number">{i + 1}</span>
          <span className="token-line__content">
            {line.map((token: Token) => {
              const key = createId();
              return <span key={key} {...getTokenProps({ token, key })} />;
            })}
          </span>
        </div>
      );
    })}
  </pre>
);

export type TcodeMarkdown = {
  inline?: boolean;
  className?: string;
  children?: ReactNode;
};
export const codeMarkdown = ({ inline, className, children }: TcodeMarkdown) => {
  const match = regexLanguage(className);
  return !inline && match ? (
    <Highlight {...defaultProps} theme={vsDark} code={String(children).replace(/\n$/, '')} language="jsx">
      {({ className: classNameChild, style, tokens, getLineProps, getTokenProps }) => (
        <HighlightWrapper
          className={classNameChild}
          style={style}
          child={children}
          getLineProps={getLineProps}
          getTokenProps={getTokenProps}
          tokens={tokens}
        />
      )}
    </Highlight>
  ) : (
    <>{children}</>
  );
};

const ReadMe = ({ markdownContent, isFetching, error, refetch, codeMarkdownFn = codeMarkdown }: TReadMe) => (
  <Loader mode={setLoaderMode({ isLoading: isFetching })}>
    <Resilience anomaly={error as Tanomaly} refetch={refetch as React.MouseEventHandler<HTMLButtonElement>}>
      <div className="markdown-body">
        <ReactMarkdown
          remarkPlugins={[remarkGfm]}
          rehypePlugins={[rehypeSlug]}
          components={{
            code: codeMarkdownFn,
          }}
        >
          {markdownContent as string}
        </ReactMarkdown>
      </div>
    </Resilience>
  </Loader>
);

export default ReadMe;
