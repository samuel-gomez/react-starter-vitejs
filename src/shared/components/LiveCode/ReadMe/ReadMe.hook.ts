import { useQuery } from '@tanstack/react-query';
import { API_URL, GITHUB_API } from 'shared/constants';
import type { Tanomaly } from 'shared/types.d';

type TuseGithubReadme = {
  githubPackage?: string;
  useQueryFn?: typeof useQuery;
};

export const useGithubReadme = ({ githubPackage, useQueryFn = useQuery }: TuseGithubReadme) => {
  const { data, isFetching, error, refetch } = useQueryFn<unknown, Tanomaly>({
    queryKey: [`${GITHUB_API}${githubPackage}/README.md`, { text: true }, API_URL.GITHUB],
    staleTime: 10000,
  });

  return { markdownContent: data, isFetching, error, refetch };
};

export type TReturnUseGithubReadme = ReturnType<typeof useGithubReadme>;
