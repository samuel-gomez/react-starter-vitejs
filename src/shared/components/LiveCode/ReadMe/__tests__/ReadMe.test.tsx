import { describe, it, expect, vi } from 'vitest';
import { render } from '@testing-library/react';
import ReadMe, { codeMarkdown, regexLanguage } from '../ReadMe';

describe('<ReadMe />', () => {
  it('Render <ReadMe /> with markdownContent, isFetching, error, refetch', () => {
    const { asFragment } = render(<ReadMe markdownContent="test" isFetching={false} error={null} refetch={vi.fn()} />);

    expect(asFragment()).toMatchSnapshot();
  });
});

describe('codeFn', () => {
  const childrenMock = ['const Button = () => <SimpleButton>Lorem Ipsum</SimpleButton>;\nexport default Button;\n'];

  it('Render codeFn with inline undefined', () => {
    const { asFragment } = render(
      codeMarkdown({
        inline: undefined,
        className: 'language-javascript',
        children: childrenMock,
      }),
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('Render codeFn with inline true', () => {
    const { asFragment } = render(
      codeMarkdown({
        inline: true,
        className: undefined,
        children: childrenMock,
      }),
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('regexLanguage', () => {
  it('Should return an array with javascript language', () => {
    const result = regexLanguage('language-javascript');
    expect(result?.length).toEqual(2);
  });

  it('Should return null when className is undefined', () => {
    const result = regexLanguage(undefined);
    expect(result).toEqual(null);
  });
});
