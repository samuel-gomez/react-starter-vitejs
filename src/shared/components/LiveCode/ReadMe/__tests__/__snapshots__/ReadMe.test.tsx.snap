// Vitest Snapshot v1

exports[`<ReadMe /> > Render <ReadMe /> with markdownContent, isFetching, error, refetch 1`] = `
<DocumentFragment>
  <div
    class="af-loader"
  >
    <div
      class="markdown-body"
    >
      <p>
        test
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`codeFn > Render codeFn with inline true 1`] = `
<DocumentFragment>
  const Button = () =&gt; &lt;SimpleButton&gt;Lorem Ipsum&lt;/SimpleButton&gt;;
export default Button;

</DocumentFragment>
`;

exports[`codeFn > Render codeFn with inline undefined 1`] = `
<DocumentFragment>
  <pre
    class="prism-code language-jsx"
    style="color: rgb(156, 220, 254); background-color: rgb(30, 30, 30); padding: 1rem; font-size: 0.8rem; position: relative;"
  >
    <button
      aria-label="Copy to clipboard"
      class="af-btn--circle af-clipboard"
      type="button"
    >
      <i
        class="glyphicon glyphicon-copy"
      />
      <span
        class="af-clipboard__message"
      >
        Copied !
      </span>
    </button>
    <div
      class="token-line"
      style="color: rgb(156, 220, 254);"
    >
      <span
        class="token-line__number"
      >
        1
      </span>
      <span
        class="token-line__content"
      >
        <span
          class="token keyword"
          style="color: rgb(86, 156, 214);"
        >
          const
        </span>
        <span
          class="token plain"
        >
           
        </span>
        <span
          class="token function-variable function maybe-class-name"
          style="color: rgb(220, 220, 170);"
        >
          Button
        </span>
        <span
          class="token plain"
        >
           
        </span>
        <span
          class="token operator"
          style="color: rgb(212, 212, 212);"
        >
          =
        </span>
        <span
          class="token plain"
        >
           
        </span>
        <span
          class="token punctuation"
          style="color: rgb(212, 212, 212);"
        >
          (
        </span>
        <span
          class="token punctuation"
          style="color: rgb(212, 212, 212);"
        >
          )
        </span>
        <span
          class="token plain"
        >
           
        </span>
        <span
          class="token arrow operator"
          style="color: rgb(212, 212, 212);"
        >
          =&gt;
        </span>
        <span
          class="token plain"
        >
           
        </span>
        <span
          class="token tag punctuation"
          style="color: rgb(212, 212, 212);"
        >
          &lt;
        </span>
        <span
          class="token tag class-name"
          style="color: rgb(78, 201, 176);"
        >
          SimpleButton
        </span>
        <span
          class="token tag punctuation"
          style="color: rgb(212, 212, 212);"
        >
          &gt;
        </span>
        <span
          class="token plain-text"
        >
          Lorem Ipsum
        </span>
        <span
          class="token tag punctuation"
          style="color: rgb(212, 212, 212);"
        >
          &lt;/
        </span>
        <span
          class="token tag class-name"
          style="color: rgb(78, 201, 176);"
        >
          SimpleButton
        </span>
        <span
          class="token tag punctuation"
          style="color: rgb(212, 212, 212);"
        >
          &gt;
        </span>
        <span
          class="token punctuation"
          style="color: rgb(212, 212, 212);"
        >
          ;
        </span>
        <span
          class="token plain"
        />
      </span>
    </div>
    <div
      class="token-line"
      style="color: rgb(156, 220, 254);"
    >
      <span
        class="token-line__number"
      >
        2
      </span>
      <span
        class="token-line__content"
      >
        <span
          class="token plain"
        />
        <span
          class="token keyword module"
          style="color: rgb(86, 156, 214);"
        >
          export
        </span>
        <span
          class="token plain"
        >
           
        </span>
        <span
          class="token keyword module"
          style="color: rgb(86, 156, 214);"
        >
          default
        </span>
        <span
          class="token plain"
        >
           
        </span>
        <span
          class="token maybe-class-name"
        >
          Button
        </span>
        <span
          class="token punctuation"
          style="color: rgb(212, 212, 212);"
        >
          ;
        </span>
      </span>
    </div>
  </pre>
</DocumentFragment>
`;
